### ISSUE TITLE
P4-B: Train code generation model with rubric rewards (GRPO)

### ISSUE THREAD

**Issue by jpequegn** [jpequegn/rubric-gates#28]
## Description — Phase B: Rubric-Rewarded Code Generation

**Goal:** Fine-tune an open-source code model using GRPO with your rubric scores as the reward signal, so it learns to generate code that inherently passes your quality gates.

**This is the core RLVR loop:** model generates code → rubric scores it → scores become rewards → model improves → generates better code.

## Why This Matters

Currently your Claude Code skills generate code, and P1-P2 evaluate it after the fact. This project flips the script: train a model that has *internalized* your quality standards. The trained model could:
- Serve as a secondary generator for simpler tasks (lower cost than Claude API)
- Power an auto-fix system that rewrites code failing the gate
- Act as a "first draft" generator that Claude then refines

### Algorithm: GRPO (Group Relative Policy Optimization)

GRPO is critic-free RL — no value model needed, which halves memory requirements. Perfect for local GPU training.

```
For each prompt:
  1. Generate K completions (group size, e.g., K=4)
  2. Score each completion with rubric engine (P1)
  3. Compute advantages: normalize scores within the group
  4. Update model: increase probability of high-scoring completions
```

### Data
- **Prompts:** Code generation tasks matching what your business users request
  - "Write a Python script that categorizes expenses by department"
  - "Create a function that parses CSV files and generates summary statistics"
  - "Build a tool that monitors a directory for new files"
- **Source:** Extract task descriptions from P3 tool registry + synthetic generation
- **Minimum:** 200+ diverse prompts, model generates 4+ completions each

### Model
- **Base:** DeepSeek-Coder-1.3B or CodeGemma-2B (fits local GPU with LoRA)
- **Fine-tuning:** LoRA + GRPO via TRL library
- **Quantization:** 4-bit base model, LoRA adapters in fp16

### Reward Function
The reward for each completion is the P1 composite score:

```python
async def rubric_reward(prompt: str, completion: str) -> float:
    """Compute reward using rubric-gates scoring engine."""
    engine = RubricEngine(config)
    result = engine.score(code=completion, filename="generated.py")
    
    # Bonus/penalty for gate tier
    gate_result = gate_evaluator.evaluate(result, completion, "generated.py")
    if gate_result.tier == GateTier.RED:
        return result.composite_score * 0.5  # Heavy penalty for red-tier
    elif gate_result.tier == GateTier.GREEN:
        return result.composite_score * 1.2  # Bonus for clean green
    
    return result.composite_score
```

### Training Config
```python
grpo_config = GRPOConfig(
    model_name="deepseek-ai/deepseek-coder-1.3b-instruct",
    group_size=4,           # Completions per prompt
    learning_rate=1e-5,
    num_epochs=3,
    max_completion_length=1024,
    kl_penalty=0.01,        # Prevent reward hacking
    lora_r=16,
    lora_alpha=32,
    batch_size=2,           # Fits on 8GB+ VRAM
)
```

### Reward Hacking Prevention
Drawing from P2's anti-gaming measures (#16):
- **KL penalty:** Prevent model from drifting too far from base
- **Length normalization:** Don't reward verbosity
- **Anti-gaming scorer:** Run P2's anti-gaming checks as part of reward
- **Diversity penalty:** Penalize repetitive completions within a group

## Interface

```python
class CodeGenTrainer(RubricTrainer):
    """Train code generation model with GRPO using rubric rewards."""
    
    def __init__(self, config: GRPOConfig, rubric_engine: RubricEngine):
        ...
    
    def prepare_prompts(self, registry: ToolCatalog | None = None) -> list[str]:
        """Generate/load training prompts from registry or synthetic source."""
    
    def compute_reward(self, prompt: str, completion: str) -> float:
        """Score completion using rubric engine + gate evaluation."""
    
    def train(self) -> TrainingResult:
        """Run GRPO training loop."""
    
    def evaluate(self, test_prompts: list[str]) -> EvalResult:
        """Evaluate trained model vs base model on rubric scores."""
    
    def export(self, path: str) -> None:
        """Export trained model for inference."""

class RubricCodeGenerator:
    """Generate code using the rubric-trained model."""
    
    def __init__(self, model_path: str):
        """Load trained model."""
    
    def generate(self, prompt: str, max_length: int = 1024) -> str:
        """Generate code for a given task description."""
    
    def generate_and_score(self, prompt: str) -> tuple[str, ScoreResult]:
        """Generate code and score it with the rubric engine."""
```

## Expected Outcomes

After training, compare base model vs trained model:

| Metric | Base Model | Trained (target) |
|--------|-----------|-------------------|
| Avg composite score | ~0.4-0.5 | > 0.7 |
| Green tier rate | ~30% | > 70% |
| Red tier rate | ~20% | < 5% |
| Security findings | Frequent | Rare |

## Acceptance Criteria

- [ ] GRPO trainer in `training/code_gen.py`
- [ ] Generator in `training/rubric_generator.py`
- [ ] Training runs on local GPU (8GB+ VRAM with LoRA + 4-bit)
- [ ] Reward function integrates P1 scorecard + P2 gate
- [ ] KL penalty and anti-gaming measures prevent reward hacking
- [ ] Trained model scores measurably higher than base model on rubric
- [ ] Training logs to W&B (or local TensorBoard)
- [ ] Tests with mock model for CI

## Dependencies

- #26 (training infrastructure)
- #9 (rubric engine — provides reward signal)
- #13 (gate evaluator — bonus/penalty in reward)
- #16 (anti-gaming — prevents reward hacking during training)
- #27 (Phase A — can optionally use distilled scorer for faster reward computation)
